src/Microsoft.AspNetCore.SignalR.Common/Internal/Encoders/Base64Encoder.cs

```diff
@@ -2,27 +2,37 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
 using System.IO;
-using System.Text;
 
 namespace Microsoft.AspNetCore.SignalR.Internal.Encoders
 {
     public class Base64Encoder : IDataEncoder
     {
-        public byte[] Decode(byte[] payload)
+        public ReadOnlySpan<byte> Decode(byte[] payload)
         {
-            var buffer = new ReadOnlyMemory<byte>(payload);
+            ReadOnlySpan<byte> buffer = payload;
             LengthPrefixedTextMessageParser.TryParseMessage(ref buffer, out var message);
 
-            return Convert.FromBase64String(Encoding.UTF8.GetString(message.ToArray()));
+            Span<byte> decoded = new byte[Base64.GetMaxDecodedFromUtf8Length(message.Length)];
+            var status = Base64.DecodeFromUtf8(message, decoded, out _, out var written);
+            Debug.Assert(status == OperationStatus.Done);
+
+            return decoded.Slice(0, written);
         }
 
         public byte[] Encode(byte[] payload)
         {
-            var buffer = Encoding.UTF8.GetBytes(Convert.ToBase64String(payload));
+            Span<byte> buffer = new byte[Base64.GetMaxEncodedToUtf8Length(payload.Length)];
+
+            var status = Base64.EncodeToUtf8(payload, buffer, out _, out var written);
+            Debug.Assert(status == OperationStatus.Done);
+
             using (var stream = new MemoryStream())
             {
-                LengthPrefixedTextMessageWriter.WriteMessage(buffer, stream);
+                LengthPrefixedTextMessageWriter.WriteMessage(buffer.Slice(0, written), stream);
```

Comment:
Similar to Decode, can we just return the `Span<byte> buffer`?

Reply:
I don't think this gets us anything - the result is eventually passed to a method the expects `byte[]`. Also, this encoder is used only in very esoteric cases which are already known to be slow (e.g. require base64 encoding). The main encoder is the [pass through encoder](https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Common/Internal/Encoders/PassThroughEncoder.cs#L13-L16) that is no op if it takes and returns the byte[].

The long term plan is to take a comprehensive look at the stack to see how we can avoid copying because now we do it like crazy all over the place. This was postponed for now until we have a functionality with good tests in place and we switch to pipelines.